solidity/contracts/AcreMultiAssetVault.sol

```diff
@@ -0,0 +1,284 @@
+    error AssetAlreadySupported();
+
+    /// @notice Reverts when the asset being deposited is not on the supported
+    ///         assets list.
+    ///         Also reverts when the asset being removed from the supported
+    ///         assets list is not on the supported assets list.
+    error AssetNotSupported();
+
+    /// @notice Reverts when a checked amount is invalid.
+    error InvalidAmount(uint256 amount);
+
+    /// @notice Reverts when the deposit owner is invalid:
+    ///         - the deposit owner is the zero address,
+    ///         - the deposit owner is the asset itself.
+    error InvalidDepositOwner(address depositOwner);
+
+    /// @notice Reverts when the receiver is invalid:
+    ///         - the receiver is the zero address,
+    ///         - the receiver is the asset itself.
+    error InvalidReceiver(address receiver);
+
+    /// @notice Reverts when a deposit is not found.
+    error DepositNotFound();
+
+    /// @notice Reverts when the amount withdrawn from the Mezo Portal doesn't
+    ///         match the deposited amount. This indicates an unexpected problem
+    ///         with the Mezo Portal integration.
+    error UnexpectedWithdrawnAmount(
+        uint256 withdrawnAmount,
+        uint256 depositedAmount
+    );
+
+    /// @notice Emitted when an asset is added to the list of supported assets.
+    event SupportedAssetAdded(address asset);
+
+    /// @notice Emitted when an asset is removed from the list of supported assets.
+    event SupportedAssetRemoved(address asset);
+
+    /// @notice Emitted when a new deposit is created.
+    event DepositCreated(
+        address indexed depositOwner,
+        address indexed asset,
+        uint256 indexed depositId,
+        uint256 amount
+    );
+
+    /// @notice Emitted when a deposit is withdrawn.
+    event DepositWithdrawn(
+        address indexed depositOwner,
+        address indexed asset,
+        uint256 indexed depositId,
+        uint256 amount,
+        address receiver
+    );
+
+    /// @notice Details of a deposit.
+    struct DepositInfo {
+        uint256 balance;
+        uint256 mezoDepositId;
+    }
+
+    /// @notice The number of deposits created. Includes the deposits that
+    ///         were fully withdrawn. This is also the identifier of the most
+    ///         recently created deposit.
+    uint256 public depositCount;
+
+    mapping(address => bool) public supportedAssets;
+
+    /// @notice Mapping of the details of deposits:
+    ///         deposit owner -> asset -> deposit id -> DepositInfo
+    /// @dev The deposit ID is unique for each deposit.
+    /// @dev When a deposit is withdrawn, the DepositInfo is deleted.
+    mapping(address => mapping(address => mapping(uint256 => DepositInfo)))
+        public deposits;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address _owner,
```

Comment:
Just flagging: Passing the `_owner` this way is a bit against the philosophy of the 2-step upgrade. We don't have any guarantee this `_owner` is correct and can manage the contract. The traditional approach with `__Ownable_init(msg.sender)` and explicit 2-step ownership transfer does not have this downside.

Reply:
Good point, I decided to introduce it after @pdyraga suggestion from https://github.com/thesis/acre/pull/762#discussion_r1795019308.

My reasoning is that we use the governance address that proved to work correctly.
What is more this is the deployment of the contract, so after we deploy the contract, we will check if the owner has been set to the correct address. If not, we will deploy another instance.